pkg/controller/validatingadmissionpolicystatus/controller.go

```diff
@@ -0,0 +1,145 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// ControllerName has "Status" in it to differentiate this controller with the other that runs in API server.
+const ControllerName = "validatingadmissionpolicy-status"
+
+// Controller is the ValidatingAdmissionPolicy Status controller that reconciles the Status field of each policy object.
+// This controller runs type checks against referred types for each policy definition.
+type Controller struct {
+	policyInformer informerv1alpha1.ValidatingAdmissionPolicyInformer
+	policyQueue    workqueue.RateLimitingInterface
+	policySynced   cache.InformerSynced
+	policyClient   admissionregistrationv1alpha1.ValidatingAdmissionPolicyInterface
+
+	// typeChecker checks the policy's expressions for type errors.
+	// Type of params is defined in policy.Spec.ParamsKind
+	// Types of object are calculated from policy.Spec.MatchingConstraints
+	typeChecker *validatingadmissionpolicy.TypeChecker
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+
+	if !cache.WaitForNamedCacheSync(ControllerName, ctx.Done(), c.policySynced) {
+		return
+	}
+
+	defer c.policyQueue.ShutDown()
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.syncPolicy, 0)
+	}
+
+	<-ctx.Done()
+}
+
+func NewController(policyInformer informerv1alpha1.ValidatingAdmissionPolicyInformer, policyClient admissionregistrationv1alpha1.ValidatingAdmissionPolicyInterface, typeChecker *validatingadmissionpolicy.TypeChecker) (*Controller, error) {
+	c := &Controller{
+		policyInformer: policyInformer,
+		policyQueue:    workqueue.NewRateLimitingQueueWithConfig(workqueue.DefaultControllerRateLimiter(), workqueue.RateLimitingQueueConfig{Name: ControllerName}),
+		policyClient:   policyClient,
+		typeChecker:    typeChecker,
+	}
+	reg, err := policyInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueuePolicy(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueuePolicy(newObj)
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	c.policySynced = reg.HasSynced
+	return c, nil
+}
+
+func (c *Controller) enqueuePolicy(policy any) {
+	if policy, ok := policy.(*v1alpha1.ValidatingAdmissionPolicy); ok {
+		// policy objects are cluster-scoped, no point include its namespace.
+		key := policy.ObjectMeta.Name
+		if key == "" {
+			utilruntime.HandleError(fmt.Errorf("cannot get name of object %v", policy))
+		}
+		c.policyQueue.Add(key)
+	}
+}
+
+func (c *Controller) syncPolicy(ctx context.Context) {
+	key, shutdown := c.policyQueue.Get()
+	if shutdown {
+		return
+	}
+	defer c.policyQueue.Done(key)
+	err := func() error {
+		key, ok := key.(string)
+		if !ok {
+			return fmt.Errorf("expect a string but got %v", key)
+		}
+		policy, err := c.policyInformer.Lister().Get(key)
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				return err
+			}
+			return c.reconcile(ctx, nil)
+		}
```

Comment:
I don't have a good sense of what it should mean to call reconcile with a nil policy argument. This would read more clearly to me:

```suggestion
		if err != nil {
			if kerrors.IsNotFound(err) {
				return nil
			}
			return err
		}
```

Reply:
L114 we cannot return here. A not found means the policy is being deleted.